sympy/diffgeom/rn.py

```diff
@@ -50,40 +54,34 @@
 ###############################################################################
 # R3
 ###############################################################################
+x, y, z, rho, psi, r, theta, phi = [Dummy(s) for s in ['x', 'y', 'z',
+                                          'rho', 'psi', 'r', 'theta', 'phi']]
 R3 = Manifold('R^3', 3)  # type: Any
 # Patch and coordinate systems.
 R3_origin = Patch('origin', R3)  # type: Any
 R3_r = CoordSystem('rectangular', R3_origin, ['x', 'y', 'z'])  # type: Any
-R3_c = CoordSystem('cylindrical', R3_origin, ['rho', 'psi', 'z'])  # type: Any
-R3_s = CoordSystem('spherical', R3_origin, ['r', 'theta', 'phi'])  # type: Any
-
-# Connecting the coordinate charts.
-x, y, z, rho, psi, r, theta, phi = [Dummy(s) for s in ['x', 'y', 'z',
-                                          'rho', 'psi', 'r', 'theta', 'phi']]
-## rectangular <-> cylindrical
-R3_r.connect_to(R3_c, [x, y, z],
-                      [sqrt(x**2 + y**2), atan2(y, x), z],
-                inverse=False, fill_in_gaps=False)
-R3_c.connect_to(R3_r, [rho, psi, z],
-                      [rho*cos(psi), rho*sin(psi), z],
-                inverse=False, fill_in_gaps=False)
-## rectangular <-> spherical
-R3_r.connect_to(R3_s, [x, y, z],
-                      [sqrt(x**2 + y**2 + z**2), acos(z/
-                            sqrt(x**2 + y**2 + z**2)), atan2(y, x)],
-                inverse=False, fill_in_gaps=False)
-R3_s.connect_to(R3_r, [r, theta, phi],
-                      [r*sin(theta)*cos(phi), r*sin(
-                          theta)*sin(phi), r*cos(theta)],
-                inverse=False, fill_in_gaps=False)
-## cylindrical <-> spherical
-R3_c.connect_to(R3_s, [rho, psi, z],
-                      [sqrt(rho**2 + z**2), acos(z/sqrt(rho**2 + z**2)), psi],
-                inverse=False, fill_in_gaps=False)
-R3_s.connect_to(R3_c, [r, theta, phi],
-                      [r*sin(theta), phi, r*cos(theta)],
-                inverse=False, fill_in_gaps=False)
-del x, y, z, rho, psi, r, theta, phi
+R3_c = CoordSystem(
+    'cylindrical', R3_origin, ['rho', 'psi', 'z'],
+    {
+        R3_r: (
```

Comment:
Does `R3_s` get linked automatically here?
One of the problem of this immutable design is that the you cannot link `R3_s` before defining.

I have an another idea as making the `transforms` dict totally external to the class, and supplied as function arguments when needed, and encourage users to build the charts outside of classes.

Reply:
In my design, `R3_s` never exists in `R3_c` nor in `R3_r`. Relation between two systems never exists in earlier-generated system. It is stored in later-generated one.  
Therefore, in this case, `R3_r` contains no relation. `R3_c` contains bidirectional relation with `R3_r`. `R3_s` contains bidirectional relations with `R3_r` and `R3_c`.

I choose this hierarchial design to include 'transform_dict` in `srepr` printing and `_hashable_content` of `CoordSystem`.  
1. `_hashable_content` needs to contain the relation. It's because two coordinate system which have same relations with reference systems are efficiently same. In other words, even if two systems have same name, patch, and notation, they are different if they have different relations with reference system.  
2. It is SymPy's rule that `eval(srepr(obj))` returns mathematically same object. If we don't include `transform_dict` in `srepr` printing, the result of `eval(srepr(_))` will not contain any transform relation, making itself completely useless. And, in order to contain `transform_dict` in `srepr` printing, it should be immutable.

I'd like to hear your thoughts on this two issues. Also, if you have idea to circumvent these, please let me know. I'm all ears!

Reply:
I see the coordinate transformation can work even if the chart is defined in only one class
But I'd suggest them to have have more symmetric structures.

Although this can be more difficult to resolve, however, I think that there can be something like `CoordSystemSymbol` defined prior to creating `CoordSystem` and the dictionary can made symmetric by taking reference to such symbols.

Reply:
My hunch of this is replacing the name of `CoordSystem` to some symbol, and tie the transformations with the name rather than `CoordSystem`
I think that the transformations dict need not even be defined bidirectional if this can be done.

```diff
diff --git a/sympy/diffgeom/diffgeom.py b/sympy/diffgeom/diffgeom.py
index 97f4cf1c34..d7af6f0567 100644
--- a/sympy/diffgeom/diffgeom.py
+++ b/sympy/diffgeom/diffgeom.py
@@ -344,11 +344,10 @@ def coord_tuple_transform_to(self, to_sys, coords):

         See the docstring of ``CoordSystem`` for examples."""
         if self != to_sys:
-
-            if to_sys in self.transform_dict:
-                transf = self.transform_dict[to_sys][0]
-            elif self in to_sys.transform_dict:
-                transf = to_sys.transform_dict[self][1]
+            if to_sys.name in self.transform_dict:
+                transf = self.transform_dict[to_sys.name][0]
+            elif self.name in to_sys.transform_dict:
+                transf = to_sys.transform_dict[self.name][1]
             else:
                 transf = self.transforms[to_sys]
             coords = transf(*coords)
diff --git a/sympy/diffgeom/rn.py b/sympy/diffgeom/rn.py
index c07605ac6b..61504457c4 100644
--- a/sympy/diffgeom/rn.py
+++ b/sympy/diffgeom/rn.py
@@ -25,11 +25,23 @@
 R2 = Manifold('R^2', 2)  # type: Any
 # Patch and coordinate systems.
 R2_origin = Patch('origin', R2)  # type: Any
-R2_r = CoordSystem('rectangular', R2_origin, ['x', 'y'])  # type: Any
+
+R2_r_sym = Dummy('R2_r_sym')
+R2_p_sym = Dummy('R2_p_sym')
+
+R2_r = CoordSystem(
+    R2_r_sym, R2_origin, ['x', 'y'],
+    {
+        R2_p_sym: (
+            Lambda((x, y), [sqrt(x**2 + y**2), atan2(y, x)]),
+            Lambda((r, theta), [r*cos(theta), r*sin(theta)])
+        )
+    }
+    )  # type: Any
 R2_p = CoordSystem(
-    'polar', R2_origin, ['r', 'theta'],
+    R2_p_sym, R2_origin, ['r', 'theta'],
     {
-        R2_r: (
+        R2_r_sym: (
             Lambda((r, theta), [r*cos(theta), r*sin(theta)]),
             Lambda((x, y), [sqrt(x**2 + y**2), atan2(y, x)])
         )
@@ -59,24 +71,45 @@
 R3 = Manifold('R^3', 3)  # type: Any
 # Patch and coordinate systems.
 R3_origin = Patch('origin', R3)  # type: Any
-R3_r = CoordSystem('rectangular', R3_origin, ['x', 'y', 'z'])  # type: Any
+
+R3_r_sym = Dummy('R3_r_sym')
+R3_c_sym = Dummy('R3_c_sym')
+R3_s_sym = Dummy('R3_s_sym')
+
+R3_r = CoordSystem(
+    R3_r_sym, R3_origin, ['x', 'y', 'z'],
+    {
+        R3_c_sym: (
+            Lambda((x, y, z), [sqrt(x**2 + y**2), atan2(y, x), z]),
+            Lambda((rho, psi, z), [rho*cos(psi), rho*sin(psi), z])
+        ),
+        R3_s_sym: (
+            Lambda((x, y, z), [sqrt(x**2 + y**2 + z**2), acos(z/sqrt(x**2 + y**2 + z**2)), atan2(y, x)]),
+            Lambda((r, theta, phi), [r*sin(theta)*cos(phi), r*sin(theta)*sin(phi), r*cos(theta)]),
+        )
+    }
+)  # type: Any
 R3_c = CoordSystem(
-    'cylindrical', R3_origin, ['rho', 'psi', 'z'],
+    R3_c_sym, R3_origin, ['rho', 'psi', 'z'],
     {
-        R3_r: (
+        R3_r_sym: (
             Lambda((rho, psi, z), [rho*cos(psi), rho*sin(psi), z]),
             Lambda((x, y, z), [sqrt(x**2 + y**2), atan2(y, x), z])
+        ),
+        R3_s_sym: (
+            Lambda((rho, psi, z), [sqrt(rho**2 + z**2), acos(z/sqrt(rho**2 + z**2)), psi]),
+            Lambda((r, theta, phi), [r*sin(theta), phi, r*cos(theta)])
         )
     }
 )  # type: Any
 R3_s = CoordSystem(
-    'spherical', R3_origin, ['r', 'theta', 'phi'],
+    R3_s_sym, R3_origin, ['r', 'theta', 'phi'],
     {
-        R3_r: (
+        R3_r_sym: (
             Lambda((r, theta, phi), [r*sin(theta)*cos(phi), r*sin(theta)*sin(phi), r*cos(theta)]),
             Lambda((x, y, z), [sqrt(x**2 + y**2 + z**2), acos(z/sqrt(x**2 + y**2 + z**2)), atan2(y, x)])
         ),
-        R3_c: (
+        R3_c_sym: (
             Lambda((r, theta, phi), [r*sin(theta), phi, r*cos(theta)]),
             Lambda((rho, psi, z), [sqrt(rho**2 + z**2), acos(z/sqrt(rho**2 + z**2)), psi])
         )
(END)
```

Reply:
Maybe it can even be better to define the names as strings and make the dictionary keys as strings.
no one may define the coordinate systems as same strings.

Reply:
@sylee957 
Coincidentally, your suggestion is exactly same with my previous attempt. I ditched that attempt because of its innate restriction:

When you define 'R2_r', you included its relation with 'R2_p'. Then, how can we define a third coordinate system and connect it to 'R2_r' without mutating it?  
My answer was we cannot. This made the entire system locked, non-expandable; which was a serious problem.

Reply:
In my case, expandable coordinate system is necessary in field of fluid mechanics. When analyzing the confined flow, one may need to define 'stretched coordinates', connecting it to existing ordinary coordinate systems.

Drawback of making the systems locked exceeds the drawback of non-symmetric relations, I reckon.

Reply:
I don't think that this disallows defining coordinate systems in non symmetric way.
If the name identifier is in either of the dictionary, it will work like your previous implementation.

Reply:
@sylee957  
Oh, I see. It will be fine if we allow asymmetric dictionaries.  
But then, wouldn't it be unnecessary to keep symmetric bi-directional dictionary if we can just make it asymmetric? It's only `transform_dict` which is asymmetric; every other features such as `coord_tuple_transform_to` are not.

Reply:
Maybe I can make it like this:

1. `R2_r` and `R2_p` know each other symmetrically and unidirectionally.  
2. If new coordinate system is introduced, it is connected to `R2_r` and `R2_p` asymmetrically and bidirectionally.

Reply:
About using string as dictionary keys: this will be problematic if one want to scrap existing system and define another with same name. 

Let's think that one imported `R2_r` from `diffgeom.rn` but want to define a new one with name `'polar'`. Since importing `R2_r` runs `diffgeom.rn` and registers `'polar'` to it, this will raise some problems. I think `Dummy` will be more suitable for the key.

Reply:
For default classes, what about like 
```
R3_r = CoordSystem(
    R3_r_sym, R3_origin, ['x', 'y', 'z'],
    {
        R3_c_sym: (Lambda((x, y, z), [sqrt(x**2 + y**2), atan2(y, x), z]),),
        R3_s_sym: (Lambda((x, y, z), [sqrt(x**2 + y**2 + z**2), acos(z/sqrt(x**2 + y**2 + z**2)), atan2(y, x)]),)
    }
)
```
so the transformations can be defined symmetrically, but without redundant edges for inverse transformations?

Reply:
@sylee957 
Yep, I thought the same.  
By the way, although I'm adopting your suggestion, I'm not sure about the advantage we'll have by making the dictionary symmetric. Could you give a lead, perhaps?

Reply:
I don't think that making the composition non symmetric is a significant issue, but I think that it was a subset of the problems of restricting a lot of freedom of how the composition graph can be formed.

However, for resolving this issue, I don't think that my above suggestion can suffice, but making the transformation dictionary not stored in the object signature could be a more valid idea.

Reply:
I have my own branch of making this possible
https://github.com/sylee957/sympy/tree/diffgeom_transform

```
R2_r = CoordSystem('rectangular', R2_origin, ['x', 'y'])
R2_p = CoordSystem('polar', R2_origin, ['r', 'theta'])
R3_r = CoordSystem('rectangular', R3_origin, ['x', 'y', 'z'])
R3_c = CoordSystem('cylindrical', R3_origin, ['rho', 'psi', 'z'],)
R3_s = CoordSystem('spherical', R3_origin, ['r', 'theta', 'phi'],)
```

And I think that it is possible to separate the transformation entirely from the coordinate systems, so they are can be manipulated independently,

However, this makes it bulkier that it should be invoked by keyword arguments
`polar.coord_tuple_transform_to(rect, [0, 2], transforms=transforms)`
But I think we can rather have a full symbolic object specialized for coordinate transformations and dispatch stuff like `coord_tuple_transform_to` and rcalls to there.

I don't think that it was hierarchically good that the objects store an another object that should be operating on top of this, as well as mutability.

Reply:
@sylee957 
Does this design allow pickling?

Reply:
@sylee957 
Maybe we can define a `dict` instance in `diffgeom` and make the other methods to use that dictionary as default. Then, as long as the user stores all transform relation in that instance, `polar.coord_tuple_transform_to(rect, [0, 2])` will be suffice for most cases.

Reply:
When will you make your pull request? Shall I merge your work into this PR?